### PR TITLE
OAuth2 Support

### PR DESCRIPTION
**Pull request by ZacharyKlein**
Adds OAuth2 dependencies (Maven & Gradle), configuration, user details mapper (extracts username and roles from OpenID JWT), and modifications to the Angular and React clients to support OAuth2/OIDC flow.

### CODE REVIEW

File: generators/client/templates/react/src/main/webapp/app/shared/reducers/authentication.ts.ejs [hunk 1]
```diff
@@ -226,7 +226,7 @@ export const logout = () => async dispatch => {
   });
 
   // fetch new csrf token
-  dispatch(getSession());
+  // dispatch(getSession());
```
Comment (JasonTypesCodes) [inline generators/client/templates/react/src/main/webapp/app/shared/reducers/authentication.ts.ejs:229]: Should this remain commented out?

File: generators/client/templates/angular/src/main/webapp/app/layouts/navbar/navbar.component.html.ejs [hunk 2]
```diff
@@ -17,7 +17,7 @@
                 </a>
             </li>
             <!-- jhipster-needle-add-element-to-menu - JHipster will add new menu items here -->
-            <li *ngSwitchCase="true" ngbDropdown class="nav-item dropdown pointer" routerLinkActive="active" [routerLinkActiveOptions]="{exact: true}">
+            <li *ngSwitchCase="true" ngbDropdown class="nav-item dropdown pointer" display="dynamic" routerLinkActive="active" [routerLinkActiveOptions]="{exact: true}">
```
Comment (mraible) [inline generators/client/templates/angular/src/main/webapp/app/layouts/navbar/navbar.component.html.ejs:20]: Why is this necessary? I'm curious.

Comment (ZacharyKlein) [reply generators/client/templates/angular/src/main/webapp/app/layouts/navbar/navbar.component.html.ejs:20]: @mraible I think it may have been a merge artifact? I was comparing with a recently-generated Boot JHipster app and those were in the diff. They aren't specifically required by the Micronaut version.

File: generators/server/templates/src/main/resources/application.yml.ejs [hunk 3]
```diff
@@ -40,17 +40,34 @@ micronaut:
         enabled: true
   security:
     enabled: true
+<%_ if (authenticationType === 'oauth2') { _%>
+    oauth2:
+      enabled: true
+      login-uri: /oauth2/authorization{/provider}
+      callback-uri: /login/oauth2/code{/provider}
```
Comment (mraible) [inline generators/server/templates/src/main/resources/application.yml.ejs:47]: This placeholder looks funny. Is the `/` supposed to be inside the placeholder?

File: generators/client/templates/react/src/main/webapp/app/shared/reducers/authentication.ts.ejs (hunk 1 above)
Comment (ZacharyKlein) [reply generators/client/templates/react/src/main/webapp/app/shared/reducers/authentication.ts.ejs:229]: @JasonTypesCodes I'm unclear as to what that line of code is supposed to do. It appears to trigger a query to get the user account, but it always fails with a 401 as the user has already been logged out.

File: generators/server/templates/src/main/resources/application.yml.ejs (hunk 3 above)
Comment (ZacharyKlein) [reply generators/server/templates/src/main/resources/application.yml.ejs:47]: Yes, that's how it is specified in the Micronaut docs.  https://micronaut-projects.github.io/micronaut-security/latest/guide/index.html#authorization-code

**Comment by mraible** [comment]
I tried creating an app with this branch tonight. When I click "sign in", I get redirected to http://localhost:8080/oauth/login/oidc and I see the following JSON in my browser.

```json
{
  "type": "/oauth/login/oidc",
  "status": 401,
  "path": "/oauth/login/oidc",
  "message": "error.http.401"
}
```

image

`npm run e2e` also fails. All CI tests pass on this branch, so I'm guessing tests need to be improved to run e2e tests.

**Comment by ZacharyKlein** [comment]
@mraible Sorry about that, apparently I missed a change to the redirect URL when I committed my changes. Please retest login when you can.

**Comment by ZacharyKlein** [comment]
@mraible The e2e tests are fixed now as well.

**Comment by mraible** [comment]
I can confirm everything works as expected. I generated a new app using the following JDL:

```
application {
  config {
    baseName blog,
    applicationType monolith,
    packageName com.jhipster.demo.blog,
    prodDatabaseType mysql,
    authenticationType oauth2,
    cacheProvider hazelcast,
    buildTool maven,
    useSass true,
    testFrameworks [protractor]
  }
  entities *
}

entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}

entity Post {
  title String required
  content TextBlob required
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{entry}
}

paginate Post, Tag with infinite-scroll
```

I put this in an `app.jdl` file in an empty directory, then ran:

```
mhipster import-jdl app.jdl
```

Once the process was complete, I ran `./mvnw` and then `npm run e2e` in another window. All tests pass! 🎉👍

File: generators/client/templates/angular/src/test/javascript/e2e/admin/administration.spec.ts.ejs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    let signInPage: SignInPage;
+
+    before(async () => {
+        await browser.get('/');
+        navBarPage = new NavBarPage(true);
+        signInPage = await navBarPage.getSignInPage();
+        <%_ if (authenticationType !== 'oauth2') { _%>
+        await signInPage.autoSignInUsing('admin', 'admin');
+        <%_ } else { _%>
+        await signInPage.loginWithOAuth('admin', 'admin');
```
Comment (mraible) [inline generators/client/templates/angular/src/test/javascript/e2e/admin/administration.spec.ts.ejs:42]: These should be changed so they can be overridden by environment variables. This was a recent change in the main generator.

```
const username = process.env.E2E_USERNAME || 'admin';
const password = process.env.E2E_PASSWORD || 'admin';
```

See https://github.com/jhipster/generator-jhipster/pull/11448 for more info.

Comment (ZacharyKlein) [reply generators/client/templates/angular/src/test/javascript/e2e/admin/administration.spec.ts.ejs]: @mraible Thanks! I made the same change in the modified tests.

**Comment by JasonTypesCodes** [comment]
https://github.com/jhipster/generator-jhipster/pull/11771 Fixes broken e2e tests in React in parent generator
